// ng-zorro-antd/esm2015/steps/steps.module.js
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NzOutletModule } from 'ng-zorro-antd/core/outlet';
import { NzIconModule } from 'ng-zorro-antd/icon';
import { NzStepComponent } from './step.component';
import { NzStepsComponent } from './steps.component';
import * as ɵngcc0 from '@angular/core';
export class NzStepsModule {
}
NzStepsModule.ɵfac = function NzStepsModule_Factory(t) { return new (t || NzStepsModule)(); };
NzStepsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NzStepsModule });
NzStepsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, NzIconModule, NzOutletModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NzStepsModule, [{
        type: NgModule,
        args: [{
                imports: [BidiModule, CommonModule, NzIconModule, NzOutletModule],
                exports: [NzStepsComponent, NzStepComponent],
                declarations: [NzStepsComponent, NzStepComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NzStepsModule, { declarations: function () { return [NzStepsComponent, NzStepComponent]; }, imports: function () { return [BidiModule, CommonModule, NzIconModule, NzOutletModule]; }, exports: function () { return [NzStepsComponent, NzStepComponent]; } }); })();
