// angular-paginator/esm2015/lib/angular-paginator.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AngularPaginatorPipe } from './pipes/angular-paginator.pipe';
import { AngularPaginatorDirective } from './directives/angular-paginator.directive';
import { AngularPaginatorService } from './services/angular-paginator.service';
import * as i0 from "@angular/core";
export class AngularPaginatorModule {
}
AngularPaginatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.1", ngImport: i0, type: AngularPaginatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AngularPaginatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.1", ngImport: i0, type: AngularPaginatorModule, declarations: [AngularPaginatorPipe, AngularPaginatorDirective], imports: [CommonModule], exports: [AngularPaginatorDirective, AngularPaginatorPipe] });
AngularPaginatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.1", ngImport: i0, type: AngularPaginatorModule, providers: [AngularPaginatorService], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.1", ngImport: i0, type: AngularPaginatorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    providers: [AngularPaginatorService],
                    declarations: [AngularPaginatorPipe, AngularPaginatorDirective],
                    exports: [AngularPaginatorDirective, AngularPaginatorPipe],
                }]
        }] });
