// ng-zorro-antd/esm2015/core/element-patch/element-patch.module.js
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NzElementPatchDirective } from './element-patch.directive';
import * as ɵngcc0 from '@angular/core';
export class NzElementPatchModule {
}
NzElementPatchModule.ɵfac = function NzElementPatchModule_Factory(t) { return new (t || NzElementPatchModule)(); };
NzElementPatchModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NzElementPatchModule });
NzElementPatchModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NzElementPatchModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [NzElementPatchDirective],
                exports: [NzElementPatchDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NzElementPatchModule, { declarations: function () { return [NzElementPatchDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NzElementPatchDirective]; } }); })();
