// ng-zorro-antd/esm2015/avatar/avatar.module.js
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { BidiModule } from '@angular/cdk/bidi';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NzIconModule } from 'ng-zorro-antd/icon';
import { NzAvatarGroupComponent } from './avatar-group.component';
import { NzAvatarComponent } from './avatar.component';
import * as ɵngcc0 from '@angular/core';
export class NzAvatarModule {
}
NzAvatarModule.ɵfac = function NzAvatarModule_Factory(t) { return new (t || NzAvatarModule)(); };
NzAvatarModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NzAvatarModule });
NzAvatarModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, NzIconModule, PlatformModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NzAvatarModule, [{
        type: NgModule,
        args: [{
                declarations: [NzAvatarComponent, NzAvatarGroupComponent],
                exports: [NzAvatarComponent, NzAvatarGroupComponent],
                imports: [BidiModule, CommonModule, NzIconModule, PlatformModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NzAvatarModule, { declarations: function () { return [NzAvatarComponent, NzAvatarGroupComponent]; }, imports: function () { return [BidiModule, CommonModule, NzIconModule, PlatformModule]; }, exports: function () { return [NzAvatarComponent, NzAvatarGroupComponent]; } }); })();
