// angular-paginator/esm2015/lib/services/angular-paginator.service.js
import { Injectable, EventEmitter } from '@angular/core';
import * as i0 from "@angular/core";
export class AngularPaginatorService {
    constructor() {
        /** instances is the collection of pagination componetns */
        this.instances = {};
        /** emits an event with instance id whenever the page number changes */
        this.change = new EventEmitter();
    }
    /**
     * creates an instance for every pipe differentiated with id
     *
     * @param instance each instance refers to a unique pagination component
     */
    registerInstance(instance) {
        if (!this.instances[instance.id]) {
            this.instances[instance.id] = instance;
            this.change.emit(instance.id);
        }
        else {
            const changed = this.updateInstance(instance);
            if (changed) {
                this.change.emit(instance.id);
            }
        }
    }
    /**
     * updates existing pagination instances available
     * and returns true if value has changed
     *
     * @param instance an unique pagination component
     */
    updateInstance(instance) {
        for (const key in instance) {
            const v1 = instance[key];
            const v2 = this.instances[instance.id][key];
            if (v1 !== v2) {
                Object.assign(this.instances[instance.id], instance);
                return true;
            }
        }
        return false;
    }
    /**
     * return instance with given id
     *
     * @param id unique id of the pagination component
     */
    getInstance(id) {
        return this.instances[id];
    }
    /**
     * return currentPage for given id
     *
     * @param id unique id of the pagination component
     */
    getCurrentPage(id) {
        return this.instances[id].currentPage;
    }
}
/** id is used to identify pagination instances */
AngularPaginatorService.id = 'ANGULAR_PAGINATOR_DEFAULT';
AngularPaginatorService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.1", ngImport: i0, type: AngularPaginatorService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
AngularPaginatorService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.0.1", ngImport: i0, type: AngularPaginatorService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.1", ngImport: i0, type: AngularPaginatorService, decorators: [{
            type: Injectable
        }] });
