// simple-sidenav/bundles/simple-sidenav.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),exports, require('@angular/animations'), require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('simple-sidenav', ['@angular/core','@angular/common','exports', '@angular/animations', '@angular/core', '@angular/common'], factory) :
    (factory(global.ng.core,global.ng.common,(global['simple-sidenav'] = {}),global.ng.animations,global.ng.core,global.ng.common));
}(this, (function (ɵngcc0,ɵngcc1,exports,animations,core,common) { 
function SimpleSidenavComponent_div_0_div_1_img_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 9);
} if (rf & 2) {
    var item_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("src", item_r2.icon, ɵngcc0.ɵɵsanitizeUrl);
} }
function SimpleSidenavComponent_div_0_div_1_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 10);
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(1, "svg", 11);
    ɵngcc0.ɵɵelement(2, "path", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("@rotate", item_r2.id === ctx_r5.activeOne.id && ctx_r5.activeOne.menu ? "rotated" : "default");
} }
var _c0 = function (a0) { return { "active": a0 }; };
function SimpleSidenavComponent_div_0_div_1_Template(rf, ctx) { if (rf & 1) {
    var _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵelementStart(1, "div", 4);
    ɵngcc0.ɵɵlistener("click", function SimpleSidenavComponent_div_0_div_1_Template_div_click_1_listener() { var restoredCtx = ɵngcc0.ɵɵrestoreView(_r9); var item_r2 = restoredCtx.$implicit; var menuIndex_r3 = restoredCtx.index; var ctx_r8 = ɵngcc0.ɵɵnextContext(2); return ctx_r8.onNavClick(item_r2, menuIndex_r3); });
    ɵngcc0.ɵɵtemplate(2, SimpleSidenavComponent_div_0_div_1_img_2_Template, 1, 1, "img", 5);
    ɵngcc0.ɵɵelementStart(3, "span", 6);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, SimpleSidenavComponent_div_0_div_1_span_5_Template, 3, 1, "span", 7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "sm-simple-sidenav", 8);
    ɵngcc0.ɵɵlistener("onSidenav", function SimpleSidenavComponent_div_0_div_1_Template_sm_simple_sidenav_onSidenav_6_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); var ctx_r10 = ɵngcc0.ɵɵnextContext(2); return ctx_r10.onSidenav.emit($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r2 = ctx.$implicit;
    var ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(10, _c0, item_r2.id === ctx_r1.activeOne.id));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r2.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r2.name);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.withArrow && item_r2.menu);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("animate", true)("activeID", ctx_r1.activeID)("animation", ctx_r1.animation)("menu", ctx_r1.activeOne.menu)("withArrow", ctx_r1.withArrow)("show", item_r2.id === ctx_r1.activeOne.id && item_r2.menu);
} }
function SimpleSidenavComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵtemplate(1, SimpleSidenavComponent_div_0_div_1_Template, 7, 12, "div", 2);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("@fadeIn", ctx_r0.animate && ctx_r0.animation && ctx_r0.animation.in)("@close", ctx_r0.animate && ctx_r0.animation && ctx_r0.animation.out);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.menu);
} }
'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var fadeIn = animations.trigger('fadeIn', [
        animations.transition('void => slide-in-stagger', [
            animations.query(':enter', animations.style({
                transform: 'translateX(-100%)',
                opacity: 0,
                height: 0,
                width: 0
            }), { optional: true }),
            animations.query(':enter', animations.stagger(55, [animations.animate("{{ duration }}ms", animations.keyframes([
                    animations.style({ transform: 'translateX(-100%)', opacity: 0, height: 0, width: 0 }),
                    animations.style({ transform: 'translateX(5%)', opacity: 0.5, height: '*' }),
                    animations.style({ transform: 'translateX(-3%)', opacity: 0.6, height: '*' }),
                    animations.style({ transform: 'translateX(2%)', opacity: 0.7, height: '*', width: '100%' }),
                    animations.style({ transform: 'translateX(0)', opacity: 1, height: '*', width: '100%' })
                ]))]), { optional: true })
        ], { params: { duration: 500 } }),
        animations.transition('* => slide-in', [
            animations.query(':self', animations.style({ height: 0 }), { optional: true }),
            animations.query(':self', animations.animate("{{ duration }}ms", animations.style({ height: '*' })), { optional: true })
        ], { params: { duration: 300 } })
    ]);
    /** @type {?} */
    var fadeOut = animations.trigger('close', [
        animations.transition('slide-out => void', [animations.animate("{{ duration }}ms", animations.style({ height: 0 }))], { params: { duration: 200 } })
    ]);
    /** @type {?} */
    var rotate = animations.trigger('rotate', [
        animations.state('default', animations.style({ transform: 'translate(0, -50%) rotate(0)' })),
        animations.state('rotated', animations.style({ transform: 'translate(0, -50%) rotate(90deg)' })),
        animations.transition('default <=> rotated', [animations.animate('100ms')])
    ]);

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SimpleSidenavComponent = /** @class */ (function () {
        function SimpleSidenavComponent() {
            this.menu = [];
            this.show = true;
            this.animate = false;
            this.withArrow = true;
            this.onSidenav = new core.EventEmitter();
            this.activeOne = {};
        }
        /**
         * @param {?} changes
         * @return {?}
         */
        SimpleSidenavComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                this.activeOne = {};
                if (changes.activeID && changes.activeID.currentValue) {
                    this.activeID = changes.activeID.currentValue;
                    this.activeID && this.menu && this.findActive();
                }
            };
        /**
         * @param {?} __0
         * @param {?} index
         * @return {?}
         */
        SimpleSidenavComponent.prototype.onNavClick = /**
         * @param {?} __0
         * @param {?} index
         * @return {?}
         */
            function (_a, index) {
                var id = _a.id, name = _a.name, icon = _a.icon;
                if (this.activeOne.id === id && this.activeOne.menu) {
                    this.activeOne = {};
                    this.onSidenav.emit(null);
                    return;
                }
                this.onSidenav.emit({ id: id, name: name, icon: icon, index: index });
                this.activeOne = { id: id, name: name, icon: icon };
                if (this.menu[index].menu) {
                    this.activeOne.menu = this.menu[index].menu;
                }
            };
        /**
         * @return {?}
         */
        SimpleSidenavComponent.prototype.findActive = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.menu.forEach(function (item) {
                    if (item.id === _this.activeID || _this.hasActive(item.menu)) {
                        _this.activeOne = item;
                        return;
                    }
                });
            };
        /**
         * @param {?} menu
         * @return {?}
         */
        SimpleSidenavComponent.prototype.hasActive = /**
         * @param {?} menu
         * @return {?}
         */
            function (menu) {
                var _this = this;
                return menu && menu.some(function (item) { return item.id === _this.activeID || _this.hasActive(item.menu); });
            };
        SimpleSidenavComponent.propDecorators = {
            menu: [{ type: core.Input }],
            show: [{ type: core.Input }],
            animation: [{ type: core.Input }],
            animate: [{ type: core.Input }],
            withArrow: [{ type: core.Input }],
            activeID: [{ type: core.Input }],
            onSidenav: [{ type: core.Output }]
        };
SimpleSidenavComponent.ɵfac = function SimpleSidenavComponent_Factory(t) { return new (t || SimpleSidenavComponent)(); };
SimpleSidenavComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SimpleSidenavComponent, selectors: [["sm-simple-sidenav"]], inputs: { menu: "menu", show: "show", animate: "animate", withArrow: "withArrow", activeID: "activeID", animation: "animation" }, outputs: { onSidenav: "onSidenav" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 1, vars: 1, consts: [["class", "simple-sidenav", 4, "ngIf"], [1, "simple-sidenav"], ["class", "simple-sidenav__item", 4, "ngFor", "ngForOf"], [1, "simple-sidenav__item"], [1, "simple-sidenav__item-content", 3, "ngClass", "click"], ["class", "simple-sidenav__icon", "alt", "icon", 3, "src", 4, "ngIf"], [1, "simple-sidenav__text"], ["class", "simple-sidenav__arrow", 4, "ngIf"], [1, "simple-sidenav-submenu", 3, "animate", "activeID", "animation", "menu", "withArrow", "show", "onSidenav"], ["alt", "icon", 1, "simple-sidenav__icon", 3, "src"], [1, "simple-sidenav__arrow"], ["focusable", "false", "viewBox", "0 0 24 24", "xmlns", "http://www.w3.org/2000/svg"], ["d", "M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"]], template: function SimpleSidenavComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, SimpleSidenavComponent_div_0_Template, 2, 3, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.show && ctx.menu && ctx.menu.length > 0);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgForOf, ɵngcc1.NgClass, SimpleSidenavComponent], styles: [".simple-sidenav{max-width:360px;overflow:hidden}.simple-sidenav__item{margin-top:3px;overflow:hidden}.simple-sidenav__item-content{position:relative;display:flex;justify-content:flex-start;align-items:center;background-color:#d6d6d6;padding:5px 15px;border-radius:4px}.simple-sidenav__item-content:hover{cursor:pointer}.simple-sidenav__icon{height:35px;margin-right:8px}.simple-sidenav__text{text-overflow:ellipsis;overflow:hidden;white-space:nowrap}.simple-sidenav__arrow{position:absolute;width:30px;height:30px;right:12px;top:50%;-webkit-transform:translate(0,-50%);transform:translate(0,-50%)}.simple-sidenav .active{background-color:#b9b9b9}.simple-sidenav-submenu .simple-sidenav{padding-left:15px}"], encapsulation: 2, data: { animation: [fadeIn, fadeOut, rotate] } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SimpleSidenavComponent, [{
        type: core.Component,
        args: [{
                selector: 'sm-simple-sidenav',
                template: "<div class=\"simple-sidenav\"\n  *ngIf=\"show && menu && menu.length > 0\"\n  [@fadeIn]=\"animate && animation && animation.in\"\n  [@close]=\"animate && animation && animation.out\">\n  <div class=\"simple-sidenav__item\" *ngFor=\"let item of menu; let menuIndex = index\">\n    <div class=\"simple-sidenav__item-content\" (click)=\"onNavClick(item, menuIndex)\" [ngClass]=\"{'active': item.id === activeOne.id}\">\n      <img class=\"simple-sidenav__icon\" [src]=\"item.icon\" alt=\"icon\" *ngIf=\"item.icon\">\n      <span class=\"simple-sidenav__text\">{{ item.name }}</span>\n      <span class=\"simple-sidenav__arrow\"\n        *ngIf=\"withArrow && item.menu\"\n        [@rotate]=\"item.id === activeOne.id && activeOne.menu ? 'rotated' : 'default'\">\n        <svg focusable=\"false\" viewBox=\"0 0 24 24\" xmlns=\"http://www.w3.org/2000/svg\"><path d=\"M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z\"></path></svg>\n      </span>\n    </div>\n    <sm-simple-sidenav\n      class=\"simple-sidenav-submenu\"\n      [animate]=\"true\"\n      [activeID]=\"activeID\"\n      [animation]=\"animation\"\n      [menu]=\"activeOne.menu\"\n      (onSidenav)=\"onSidenav.emit($event)\"\n      [withArrow]=\"withArrow\"\n      [show]=\"item.id === activeOne.id && item.menu\">\n    </sm-simple-sidenav>\n  </div>\n</div>\n",
                encapsulation: core.ViewEncapsulation.None,
                animations: [fadeIn, fadeOut, rotate],
                styles: [".simple-sidenav{max-width:360px;overflow:hidden}.simple-sidenav__item{margin-top:3px;overflow:hidden}.simple-sidenav__item-content{position:relative;display:flex;justify-content:flex-start;align-items:center;background-color:#d6d6d6;padding:5px 15px;border-radius:4px}.simple-sidenav__item-content:hover{cursor:pointer}.simple-sidenav__icon{height:35px;margin-right:8px}.simple-sidenav__text{text-overflow:ellipsis;overflow:hidden;white-space:nowrap}.simple-sidenav__arrow{position:absolute;width:30px;height:30px;right:12px;top:50%;-webkit-transform:translate(0,-50%);transform:translate(0,-50%)}.simple-sidenav .active{background-color:#b9b9b9}.simple-sidenav-submenu .simple-sidenav{padding-left:15px}"]
            }]
    }], function () { return []; }, { menu: [{
            type: core.Input
        }], show: [{
            type: core.Input
        }], animate: [{
            type: core.Input
        }], withArrow: [{
            type: core.Input
        }], onSidenav: [{
            type: core.Output
        }], activeID: [{
            type: core.Input
        }], animation: [{
            type: core.Input
        }] }); })();
        return SimpleSidenavComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SimpleSidenavModule = /** @class */ (function () {
        function SimpleSidenavModule() {
        }
SimpleSidenavModule.ɵfac = function SimpleSidenavModule_Factory(t) { return new (t || SimpleSidenavModule)(); };
SimpleSidenavModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SimpleSidenavModule });
SimpleSidenavModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            common.CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SimpleSidenavModule, [{
        type: core.NgModule,
        args: [{
                declarations: [SimpleSidenavComponent],
                imports: [
                    common.CommonModule
                ],
                exports: [SimpleSidenavComponent]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SimpleSidenavModule, { declarations: [SimpleSidenavComponent], imports: [ɵngcc1.CommonModule], exports: [SimpleSidenavComponent] }); })();
        return SimpleSidenavModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.SimpleSidenavComponent = SimpleSidenavComponent;
    exports.SimpleSidenavModule = SimpleSidenavModule;
    exports.ɵa = fadeIn;
    exports.ɵb = fadeOut;
    exports.ɵc = rotate;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
