// simple-sidenav/esm2015/lib/simple-sidenav.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { SimpleSidenavComponent } from './simple-sidenav.component';
import { CommonModule } from '@angular/common';
import * as ɵngcc0 from '@angular/core';
export class SimpleSidenavModule {
}
SimpleSidenavModule.ɵfac = function SimpleSidenavModule_Factory(t) { return new (t || SimpleSidenavModule)(); };
SimpleSidenavModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SimpleSidenavModule });
SimpleSidenavModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SimpleSidenavModule, [{
        type: NgModule,
        args: [{
                declarations: [SimpleSidenavComponent],
                imports: [
                    CommonModule
                ],
                exports: [SimpleSidenavComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SimpleSidenavModule, { declarations: function () { return [SimpleSidenavComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [SimpleSidenavComponent]; } }); })();
