// angular-paginator/esm2015/lib/pipes/angular-paginator.pipe.js
import { Pipe } from '@angular/core';
import { AngularPaginatorService } from '../services/angular-paginator.service';
import * as i0 from "@angular/core";
import * as i1 from "../services/angular-paginator.service";
/**
 * Pagination Pipe which paginates the array for the provided arguments such as `itemsPerPage` and `currentPage`
 */
export class AngularPaginatorPipe {
    /**
     *
     * @param angularPaginatorService service for angular pagiantor
     */
    constructor(angularPaginatorService) {
        this.angularPaginatorService = angularPaginatorService;
    }
    /**
     * this pipes paginates the array for the provided arguments such as `itemsPerPage` and `currentPage`
     *
     * @param array input array for which the manipulation happens
     * @param args input arguments for the paginator pipe
     */
    transform(array, args) {
        const instance = this.createInstance(array, args);
        // create pagination information
        this.angularPaginatorService.registerInstance(instance);
        // set the slicing range
        const start = (instance.currentPage - 1) * instance.itemsPerPage;
        const end = instance.currentPage * instance.itemsPerPage;
        return array.slice(start, end);
    }
    /**
     * create an instance for each pipe
     *
     * @param array input array for which the manipulation happens
     * @param args input arguments for the paginator pipe
     */
    createInstance(array, args) {
        return {
            id: (args === null || args === void 0 ? void 0 : args.id) ? args.id : AngularPaginatorService.id,
            currentPage: (args === null || args === void 0 ? void 0 : args.currentPage) ? args.currentPage : 1,
            itemsPerPage: (args === null || args === void 0 ? void 0 : args.itemsPerPage) ? args.itemsPerPage : 10,
            totalItems: array.length,
        };
    }
}
AngularPaginatorPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.1", ngImport: i0, type: AngularPaginatorPipe, deps: [{ token: i1.AngularPaginatorService }], target: i0.ɵɵFactoryTarget.Pipe });
AngularPaginatorPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.0.1", ngImport: i0, type: AngularPaginatorPipe, name: "angularPaginator" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.1", ngImport: i0, type: AngularPaginatorPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'angularPaginator',
                }]
        }], ctorParameters: function () { return [{ type: i1.AngularPaginatorService }]; } });
